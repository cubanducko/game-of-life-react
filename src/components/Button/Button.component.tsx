import styled, { css } from "styled-components";

export type ButtonProps = {
  variant?: ButtonVariants;
};

export enum ButtonVariants {
  PRIMARY = "primary",
  SECONDARY = "secondary",
}

export const Button = styled.button<ButtonProps>`
  position: relative;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  border-radius: ${({ theme }) => theme.spacing(0.5)};
  font-weight: 700;
  height: ${({ theme }) => theme.spacing(5)};
  min-width: ${({ theme }) => theme.spacing(10)};
  cursor: pointer;

  ${({ variant = ButtonVariants.PRIMARY }) =>
    variant === ButtonVariants.PRIMARY
      ? primaryButtonStyles
      : secondaryButtonStyles};
`;

const primaryButtonStyles = css`
  border: 1px solid transparent;
  background: ${({ theme }) => theme.palette.primary.main};
  color: white;
  :hover,
  :active {
    background: ${({ theme }) => theme.palette.primary.accent};
  }
`;

const secondaryButtonStyles = css`
  border: 1px solid ${({ theme }) => theme.palette.primary.main};
  background: transparent;
  color: ${({ theme }) => theme.palette.primary.main};

  :hover,
  :active {
    background: ${({ theme }) => theme.palette.primary.accent};
    color: white;
  }
`;
