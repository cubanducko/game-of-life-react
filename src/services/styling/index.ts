export * from "./spacing";
export * from "./breakpoints";
export * from "./theme";
export * from "./GlobalStyles.component";
