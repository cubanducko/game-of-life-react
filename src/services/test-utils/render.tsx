import {
  render as rtlRender,
  RenderOptions,
  RenderResult,
} from "@testing-library/react";
import { FunctionComponent, PropsWithChildren, ReactElement } from "react";
import { createTheme, Theme } from "services/styling";
import { ThemeProvider } from "styled-components";

const wrapper = ({
  theme: themeVariables,
}: IWrapperOptions): FunctionComponent => {
  const RTLWrapper = ({
    children,
  }: PropsWithChildren<unknown>): ReactElement => {
    const theme = createTheme(themeVariables);

    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
  };

  return RTLWrapper;
};

export interface IWrapperOptions {
  theme?: Theme;
}

const renderWithProviders = (
  element: ReactElement,
  options: IWrapperOptions = {}
): RenderResult => {
  return rtlRender(element, {
    wrapper: wrapper(options),
  } as RenderOptions);
};

export { default as userEvent } from "@testing-library/user-event";
export * from "@testing-library/react";
export { renderWithProviders as render };
