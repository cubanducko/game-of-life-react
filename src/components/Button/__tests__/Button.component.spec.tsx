import { render } from "services/test-utils";
import { Button, ButtonVariants } from "../Button.component";

describe("Button.component", () => {
  it("renders correctly the primary variant", () => {
    const { container } = render(<Button>Example button</Button>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders correctly the secondary variant", () => {
    const { container } = render(
      <Button variant={ButtonVariants.SECONDARY}>Example button</Button>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
