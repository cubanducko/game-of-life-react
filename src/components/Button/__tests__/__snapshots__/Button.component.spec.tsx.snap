// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.component renders correctly the primary variant 1`] = `
.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 4px;
  font-weight: 700;
  height: 40px;
  min-width: 80px;
  cursor: pointer;
  border: 1px solid transparent;
  background: rgb(0,85,255);
  color: white;
}

.c0:hover,
.c0:active {
  background: rgb(0,72,217);
}

<button
  class="c0"
>
  Example button
</button>
`;

exports[`Button.component renders correctly the secondary variant 1`] = `
.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 4px;
  font-weight: 700;
  height: 40px;
  min-width: 80px;
  cursor: pointer;
  border: 1px solid rgb(0,85,255);
  background: transparent;
  color: rgb(0,85,255);
}

.c0:hover,
.c0:active {
  background: rgb(0,72,217);
  color: white;
}

<button
  class="c0"
>
  Example button
</button>
`;
